zanata-war/src/main/java/org/zanata/action/ProjectPermissionDialog.java

```diff
@@ -0,0 +1,250 @@
+            if (membership.getPerson().equals(person)) {
+                localeRoles.put(membership.getLocale(), membership.getRole());
+            }
+        }
+
+        data =
+            new PersonProjectMemberships(person, projectRoles, localeRoles);
+        LocaleService localeServiceImpl = ServiceLocator.instance().getInstance(
+                LocaleServiceImpl.class);
+        List<HLocale> locales =
+                localeServiceImpl.getSupportedLanguageByProject(getProject().getSlug());
+        data.ensureLocalesPresent(locales);
+    }
+
+    /**
+     * Ensure that the permission dialog has no person selected.
+     */
+    public void clearPerson() {
+        log.info("clear person from {}", this);
+        data = null;
+    }
+
+    /**
+     * Indicate whether any of the permissions are selected in the person
+     * permission dialog.
+     *
+     * @return false if no permissions are selected in the dialog, otherwise true
+     */
+    public boolean anySelected() {
+        return data != null && data.hasAnyPermissions();
+    }
+
+    /**
+     * Update role membership for a project-specific role based on the current
+     * checkbox value.
+     *
+     * @param role the role to assign or remove for the current person
+     * @param checked the current checkbox value
+     */
+    public void bindProjectRole(String role, boolean checked) {
+        if(StringUtils.equalsIgnoreCase(role, ProjectRole.TranslationMaintainer.name())) {
+            data.setTranslationMaintainer(checked);
+        } else if(StringUtils.equalsIgnoreCase(role, ProjectRole.Maintainer.name())) {
+            data.setMaintainer(checked);
+        }
+    }
+
+    /**
+     * Update role membership for a locale-specific role based on the current
+     * checkbox value.
+     *
+     * @param localeRole represents both the locale and the role
+     *                   format: {localeId}:{role}. e.g en-US:Reviewer
+     * @param checked the current checkbox value
+     */
+    public void bindTranslationRole(String localeRole, boolean checked) {
+        String[] localeRoleList = StringUtils.split(localeRole, ':');
+        HLocale hLocale = localeServiceImpl.getByLocaleId(localeRoleList[0]);
+        String role = localeRoleList[1];
+
+        for (PersonProjectMemberships.LocaleRoles localeRoles: data.getLocaleRoles()) {
+            if(localeRoles.getLocale().equals(hLocale)) {
+                if (StringUtils.equalsIgnoreCase(role, LocaleRole.Translator.name())) {
+                    localeRoles.setTranslator(checked);
+                } else if (StringUtils.equalsIgnoreCase(role, LocaleRole.Reviewer.name())) {
+                    localeRoles.setReviewer(checked);
+                } else if (StringUtils.equalsIgnoreCase(role, LocaleRole.Coordinator.name())) {
+                    localeRoles.setCoordinator(checked);
+                }
+                return;
+            }
+        }
```

Comment:
I think it will be more readable to use Iterables.tryFind(data.getLocaleRoles(), <predicate to test hLocale>) first and then base on the optional LocaleRoles, check things.